In [1]:
import numpy as np
import random

# Exercise 1

## Step 1.1

In [2]:
#Create a list of 1-13.
deck1Set = [x for x in range(1,14)]
#Since we do not consider about the faces, just simply copy the list for 4 times.
deck = deck1Set*4
print(deck)

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13]


## Step1.2

In [3]:
#Sum up the values.
maximum=sum(deck)
print(maximum)

364


# Exercise 2

## Step 2.1

In [4]:
#Create the tuples of players' cards on hand.
player1=();
player2=();
#Define the function. 
def deal():
    np.random.shuffle(deck)
    global player1
    global player2
#Give the value to the global variables.
    player1 =tuple (deck[:26]);
    player2 =tuple (deck[26:]);
#Print the result we want.
    return("Player1: "+str(player1)+"\nPlayer2: "+str(player2))

## Step2.2

In [5]:
#Print out the cards for two players.
print(deal())
#Calculate the number of the cards and the sum value.
print('Player1 has '+str(len(player1))+" cards, and the total value is "+ str(sum(player1)))
print('Player2 has '+str(len(player2))+" cards, and the total value is "+ str(sum(player2)))


Player1: (6, 5, 5, 13, 12, 8, 3, 12, 3, 1, 2, 2, 5, 9, 13, 4, 6, 1, 9, 12, 8, 2, 4, 7, 4, 13)
Player2: (8, 10, 2, 1, 7, 9, 10, 12, 10, 11, 9, 1, 3, 10, 11, 5, 13, 4, 7, 11, 3, 11, 6, 6, 7, 8)
Player1 has 26 cards, and the total value is 169
Player2 has 26 cards, and the total value is 195


# Exercise 3

## Step3.1

1. Shuffle deck
2. Deal half of deck to each player
3. Each player choose two random cards from their tuple list.
4. Compare the sum of the pairs.
5. Winner take all four cards and score will increase the sum of the 4 cards.
6. There is a tie, both players give one more random card in the round.(If still a tie, add one more.)
7. Keep going.
8. When players ran out of the cards during a tie or just have one card left, return the cards to their own pile and add up the values.
9. Game terminates after a maximum of 13 rounds (occurs when there are no ties) and the deck is exhausted. 
10. The player with the highest total card points (in their discard pile) wins the game.                                   

## Step3.2

In [6]:
def play():
    #Define the numbers that I will use in the output.
    p1Score=p2Score=roundNum=0
    #Run function deal first to get the random cards for players.
    deal()
    global player1
    global player2
    #The previous section required tuples, now I want to switch to list first to modify it.
    player1=list(player1)
    player2=list(player2)
    #Checking the game process by printing.
    #print(player1)
    #print(player2)
    #Start a loop for the whole game process.
    while player1!=[]: 
        #What if only one card left after a round-return the values.
        if len(player1)==1 or len(player2)==1:
            p1Score = p1Score + sum(player1)
            p2Score = p2Score + sum(player2)
            break
        #Choose 2 cards randomly from the list.        
        player1_round = random.sample(player1, 2)
        player2_round = random.sample(player2, 2)
        #print(player1_round,player2_round)
        #Delete the chosen cards so they will not be used twice.
        player1.remove(player1_round[0])
        player1.remove(player1_round[1])
        player2.remove(player2_round[0])
        player2.remove(player2_round[1])
        #Checking the game process by printing.
        #print(player1)
        #print(player2)
        #What if two cards have the same sum - tie.
        while sum(player1_round)==sum(player2_round):
            #If these two are the last two cards-tie-return the values. 
            if player1==[] or player2==[]:
                break
            #If not, draw another card and add the sum together and then compare.
            else:
                p1New=random.choice(player1)
                p2New=random.choice(player2)
                player1_round.append(p1New)
                #Checking the game process by printing.
                #print('tie warning: draw another card')
                #print(player1_round)
                player2_round.append(p2New)
                #print(player2_round)
                player1.remove(p1New)
                player2.remove(p2New)
            #After all these, if they run out of the cards, return the values.    
            if player1==[] or player2==[]:
                break
        
        #After have two sets of cards with different sum, then compare and add the score to players. 
        if sum(player1_round)>sum(player2_round):
                p1Score = p1Score + sum(player1_round) + sum(player2_round)
        elif sum(player1_round)<sum(player2_round): 
                p2Score = p2Score + sum(player1_round) + sum(player2_round)
        else:
                p1Score = p1Score + sum(player1_round)
                p2Score = p2Score + sum(player2_round)
        # Count the round number.   
        roundNum=roundNum+1
    #Return the coresponding results.
    if p1Score>p2Score:
        return ('The game lasts for '+ str(roundNum)+' rounds. \nPlayer1 is the winner and the score is: '+str(p1Score)+'\nPlayer2 score is: '+str(p2Score))
    elif p1Score<p2Score:
        return ('The game lasts for '+ str(roundNum)+' rounds. \nPlayer2 is the winner and the score is: '+str(p2Score)+'\nPlayer1 score is: '+str(p1Score))
    else:
        return ('The game lasts for '+ str(roundNum)+' rounds. \nThere is a tie and the score is: '+str(p1Score))
    
    

## Step3.3

In [7]:
print(play())

The game lasts for 13 rounds. 
Player1 is the winner and the score is: 230
Player2 score is: 134


# Exercise 4

In [8]:
result=[]
def playNew():
    #Define the numbers that I will use in the output.
    p1Score=p2Score=roundNum=0
    #Run function deal first to get the random cards for players.
    deal()
    global player1
    global player2
    player1Initial=sum(player1)
    player2Initial=sum(player2)
    #The previous section required tuples, now I want to switch to list first to modify it.
    player1=list(player1)
    player2=list(player2)
    #Checking the game process by printing.
    #print(player1)
    #print(player2)
    #Start a loop for the whole game process.
    while player1!=[]: 
        #What if only one card left after a round-return the values.
        if len(player1)==1 or len(player2)==1:
            p1Score = p1Score + sum(player1)
            p2Score = p2Score + sum(player2)
            break
        #Choose 2 cards randomly from the list.        
        player1_round = random.sample(player1, 2)
        player2_round = random.sample(player2, 2)
        #print(player1_round,player2_round)
        #Delete the chosen cards so they will not be used twice.
        player1.remove(player1_round[0])
        player1.remove(player1_round[1])
        player2.remove(player2_round[0])
        player2.remove(player2_round[1])
        #Checking the game process by printing.
        #print(player1)
        #print(player2)
        #What if two cards have the same sum - tie.
        while sum(player1_round)==sum(player2_round):
            #If these two are the last two cards-tie-return the values. 
            if player1==[] or player2==[]:
                break
            #If not, draw another card and add the sum together and then compare.
            else:
                p1New=random.choice(player1)
                p2New=random.choice(player2)
                player1_round.append(p1New)
                #Checking the game process by printing.
                #print('tie warning: draw another card')
                #print(player1_round)
                player2_round.append(p2New)
                #print(player2_round)
                player1.remove(p1New)
                player2.remove(p2New)
            #After all these, if they run out of the cards, return the values.    
            if player1==[] or player2==[]:
                break
        
        #After have two sets of cards with different sum, then compare and add the score to players. 
        if sum(player1_round)>sum(player2_round):
                p1Score = p1Score + sum(player1_round) + sum(player2_round)
        elif sum(player1_round)<sum(player2_round): 
                p2Score = p2Score + sum(player1_round) + sum(player2_round)
        else:
                p1Score = p1Score + sum(player1_round)
                p2Score = p2Score + sum(player2_round)
        # Count the round number.   
        roundNum=roundNum+1
    #Return the coresponding results.
    global result
    if p1Score>p2Score:
        result=[player1Initial,player2Initial,p1Score,p2Score]
        return result
    elif p1Score<p2Score:
        result=[player2Initial,player2Initial,p2Score,p1Score]
        return result
    else:
        return result

winnerInitial=[]
winnerFinal=[]
loserInitial=[]
loserFinal=[]

LowInitialWinnerCount=0
for i in range(20):
    playNew()
    winnerInitial.append(result[0])
    winnerFinal.append(result[2])
    loserInitial.append(result[1])
    loserFinal.append(result[3])
    if result[0]<result[1]:
        LowInitialWinnerCount=LowInitialWinnerCount+1

print('winnerInitial:'+str(winnerInitial))
print('winnerFinal:'+str(winnerFinal))
print('loserInitial:'+str(loserInitial))
print('loserFinal:'+str(loserFinal))

    
    


winnerInitial:[195, 198, 185, 184, 192, 193, 184, 189, 188, 180, 175, 183, 191, 209, 182, 179, 195, 189, 185, 187]
winnerFinal:[282, 261, 184, 227, 189, 201, 218, 232, 201, 209, 196, 235, 235, 286, 280, 202, 209, 214, 207, 194]
loserInitial:[195, 166, 179, 180, 172, 193, 184, 189, 176, 180, 175, 181, 191, 209, 182, 179, 195, 189, 185, 187]
loserFinal:[82, 103, 180, 137, 175, 163, 146, 132, 163, 155, 168, 129, 129, 78, 84, 162, 155, 150, 157, 170]


## Step4.1

In [9]:
#Define function mean:
def mean(lst): 
    return sum(lst) / len(lst) 
#For winner data:
print('Winners final scores:'+str(winnerFinal)+'\nWinner minimum:'+str(min(winnerFinal))+'\nWinner maximum:'+str(max(winnerFinal))+'\nWinner mean:'+str(mean(winnerFinal)))
#For loser data:
print('Losers final scores:'+str(loserFinal)+'\nLoser minimum:'+str(min(loserFinal))+'\nLoser maximum:'+str(max(loserFinal))+'\nLoser mean:'+str(mean(loserFinal)))

Winners final scores:[282, 261, 184, 227, 189, 201, 218, 232, 201, 209, 196, 235, 235, 286, 280, 202, 209, 214, 207, 194]
Winner minimum:184
Winner maximum:286
Winner mean:223.1
Losers final scores:[82, 103, 180, 137, 175, 163, 146, 132, 163, 155, 168, 129, 129, 78, 84, 162, 155, 150, 157, 170]
Loser minimum:78
Loser maximum:180
Loser mean:140.9


Since the sums of the winner scores and the loser scores in each game should equal to the total value of the deck(which equal to 364 ), both winner's minimum and maximum are over the half(364/2=182). 
Similarly, both loser's minimum and maximum are less than the half(364/2=182). 
The winner's minimum plus loser's maximum is equal to 364, also, the loser's minimum plus winner's maximum is equal to 364. 
Last, sum of two means also equal to 364.

## Step4.2

In [10]:
#Use the counting variable from the previous loop to calculate the proportion.
ProportionWinWithLowInitial=LowInitialWinnerCount/20
print('The proportion of games resulted in a winner that had a lower initial total than their opponent: '+ str(ProportionWinWithLowInitial)+'\nThe lowest initial total to win a game:'+str(min(winnerInitial))+'\nthe highest initial total to lose a game: '+str(max(loserInitial)))

The proportion of games resulted in a winner that had a lower initial total than their opponent: 0.0
The lowest initial total to win a game:175
the highest initial total to lose a game: 209


## Step4.3

In [11]:
#For winners:
winnerCo=np.corrcoef(winnerInitial, winnerFinal)[0, 1]
#For losers:
loserCo=np.corrcoef(loserInitial, loserFinal)[0, 1]
print('The correlation between the initial and final totals for winners: '+ str(winnerCo)+'\nThe correlation between the initial and final totals for losers: '+ str(loserCo))

The correlation between the initial and final totals for winners: 0.4905434585736774
The correlation between the initial and final totals for losers: -0.37824707643965594


I ran the playNew() function several time, which gave me a big sample size, then I found out the correlation between initial scores and final scores was not signaficant. I saw both positive and negative coefficients(most positive tho). The range of the absolute values of the coefficients is also wide, with a great variation.